// @walletconnect/client/dist/cjs/constants/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLIENT_STORAGE_OPTIONS = exports.CLIENT_EVENTS = exports.CLIENT_SHORT_TIMEOUT = exports.CLIENT_DEFAULT = exports.CLIENT_CONTEXT = void 0;
exports.CLIENT_CONTEXT = "client";
exports.CLIENT_DEFAULT = {
    name: exports.CLIENT_CONTEXT,
    logger: "error",
    controller: false,
    relayUrl: "wss://relay.walletconnect.com",
};
exports.CLIENT_SHORT_TIMEOUT = 50;
exports.CLIENT_EVENTS = {
    pairing: {
        proposal: "pairing_proposal",
        updated: "pairing_updated",
        upgraded: "pairing_upgraded",
        extended: "pairing_extended",
        created: "pairing_created",
        deleted: "pairing_deleted",
        sync: "pairing_sync",
    },
    session: {
        proposal: "session_proposal",
        updated: "session_updated",
        upgraded: "session_upgraded",
        extended: "session_extended",
        created: "session_created",
        deleted: "session_deleted",
        notification: "session_notification",
        request: "session_request",
        response: "session_response",
        sync: "session_sync",
    },
};
exports.CLIENT_STORAGE_OPTIONS = {
    database: ":memory:",
};
