// @walletconnect/client/dist/cjs/constants/relayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RELAYER_STORAGE_OPTIONS = exports.RELAYER_RECONNECT_TIMEOUT = exports.RELAYER_PROVIDER_EVENTS = exports.RELAYER_SUBSCRIBER_SUFFIX = exports.RELAYER_EVENTS = exports.RELAYER_CONTEXT = exports.RELAYER_DEFAULT_RELAY_URL = exports.RELAYER_DEFAULT_LOGGER = exports.RELAYER_DEFAULT_PROTOCOL = void 0;
const time_1 = require("./time");
exports.RELAYER_DEFAULT_PROTOCOL = "waku";
exports.RELAYER_DEFAULT_LOGGER = "error";
exports.RELAYER_DEFAULT_RELAY_URL = "wss://relay.walletconnect.com";
exports.RELAYER_CONTEXT = "relayer";
exports.RELAYER_EVENTS = {
    payload: "relayer_payload",
    connect: "relayer_connect",
    disconnect: "relayer_disconnect",
    error: "relayer_error",
};
exports.RELAYER_SUBSCRIBER_SUFFIX = "_subscription";
exports.RELAYER_PROVIDER_EVENTS = {
    payload: "payload",
    connect: "connect",
    disconnect: "disconnect",
    error: "error",
};
exports.RELAYER_RECONNECT_TIMEOUT = time_1.ONE_SECOND;
exports.RELAYER_STORAGE_OPTIONS = {
    database: ":memory:",
};
